lib/core/bech32_cosmos.dart

```diff
@@ -0,0 +1,358 @@
+    -1,
+    -1,
+    -1
+  ];
+
+  /// Find the polynomial with value coefficients mod the generator as 30-bit.
+  static int _polymod(Int8List values) {
+    var c = 1;
+    values.forEach((v_i) {
+      var c0 = (c >> 25) & 0xff;
+      c = c & 0x1ffffff << 5 ^ ((v_i) & 0xff);
+      if (c0 & 1 != 0) c = c ^ 0x3b6a57b2;
+      if (c0 & 2 != 0) c = c ^ 0x26508e6d;
+      if (c0 & 4 != 0) c = c ^ 0x1ea119fa;
+      if (c0 & 8 != 0) c = c ^ 0x3d4233dd;
+      if (c0 & 16 != 0) c = c ^ 0x2a1462b3;
+    });
+    return c;
+  }
+
+  /// Expand a HRP for use in checksum computation.
+  static Int8List _expandHrp(String hrp) {
+    var hrpLength = hrp.length;
+    var ret = Int8List(hrpLength * 2 + 1);
+    for (var i = 0; i < hrpLength; i++) {
+      var c = hrp.codeUnits[i] & 0x7f; // Limit to standard 7-bit ASCII
+      ret[i] = ((c >> 5) & 0x07);
+      ret[i + hrpLength + 1] = (c & 0x1f);
+    }
+    ret[hrpLength] = 0;
+    return ret;
+  }
+
+  /// Verify a checksum.
+  static bool _verifyChecksum(String hrp, Int8List values) {
+    var hrpExpanded = _expandHrp(hrp);
+    var combined = Int8List(hrpExpanded.length + values.length);
+    combined.setRange(0, hrpExpanded.length, hrpExpanded);
+    combined.setRange(
+        hrpExpanded.length, values.length + hrpExpanded.length, values);
+    return _polymod(combined) == 1;
+  }
+
+  /// Create a checksum.
+  static Int8List _createChecksum(String hrp, Int8List values) {
+    var hrpExpanded = _expandHrp(hrp);
+    var enc = Int8List(hrpExpanded.length + values.length + 6);
+    enc.setRange(0, hrpExpanded.length, hrpExpanded);
+    enc.setRange(
+        hrpExpanded.length, values.length + hrpExpanded.length, values);
+
+    var mod = _polymod(enc) ^ 1;
+    var ret = Int8List(6);
+    for (var i = 0; i < 6; i++) {
+      ret[i] = ((mod >> (5 * (5 - i))) & 31);
+    }
+    return ret;
+  }
+
+  /// Encodes a Bech32 string.
+  static String encodeData(Bech32Data bech32) {
+    return encode(bech32.hrp, bech32.data);
+  }
+
+  /// Encodes a Bech32 string.
+  static String encode(String humanReadablePart, Int8List data) {
+    var hrp = humanReadablePart;
+
+    if (hrp.isEmpty) {
+      throw AddressFormatExceptionHrpLength('Human-readable part is too short');
+    }
+    if (hrp.length > 83) {
+      throw AddressFormatExceptionHrpLength('Human-readable part is too long');
+    }
+
+    hrp = hrp.toLowerCase();
+    var checksum = _createChecksum(hrp, data);
+    var combined = Int8List(data.length + checksum.length);
+    combined.setRange(0, data.length, data);
+    combined.setRange(data.length, data.length + checksum.length, checksum);
+
+    var sb = StringBuffer(hrp.length + 1 + combined.length);
+    sb.write(hrp);
+    sb.write('1');
+    combined.forEach((b) {
+      sb.write(CHARSET[b.toInt()]);
+    });
+    return sb.toString();
+  }
+
+  /// Decodes a Bech32 string.
+  static Bech32Data decode(String str) {
+    var lower = false;
+    var upper = false;
+    if (str.length < 8) {
```

Comment:
the numbers 8, 90 can be probably be replaced with constant variables.
as well as 33, 126. I have seen good constant use everywhere else, may be this will be cleaned up in another pass.

Reply:
you're spot on w/ that being weird/bad, but Bech32Cosmos is a verbatim port of the kotlin implementation and as of yet isn't sufficiently tested, so i erred on the side of keeping it more similar to the kotlin for the sake of enabling easy side-by-side comparison, until we're absolutely confident the dart implementation isn't broken in some way.

Reply:
It's approved per your explanation.